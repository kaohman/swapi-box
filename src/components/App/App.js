import React, { Component } from 'react';
import '../../normalize.css';
import './App.css';
import '../../main.scss';
import ControlForm from '../ControlForm/ControlForm';
import CardContainer from '../CardContainer/CardContainer';
import Landing from '../Landing/Landing';
import API from '../api/api';

class App extends Component {
  constructor() {
    super();
    this.state = {
      currentPage: 'landing',
      loaded: false,
      scrollText: {},
      favorites: [],
      people: [],
      planets: [],
      vehicles: []
    }
  }

  getData = async (item) => {
    if (this.state[item].length === 0) {
      const results = await API.fetchData(`https://swapi.co/api/${item}`);
      // const allData = await this.getNextPageData(results);
      const unresolvedPromises = await this.getPeopleData(results.results);
      const finalData = await Promise.all(unresolvedPromises);
      this.setState({
        [item]: finalData,
        currentPage: item
      });
    } 
  }

  getNextPageData = async (prevResults) => {
    let allResults = prevResults.results;
    do {
      const results = await API.fetchData(prevResults.next);
      allResults = await allResults.concat(results.results);
      prevResults = results;
    } while (prevResults.next);
    return allResults
  }

  getPeopleData = (data) => {
    return data.map(async person => {
      const homeResult = await API.fetchData(person.homeworld);
      const homeworld = [homeResult.name, homeResult.population];
      const speciesResult = await API.fetchData(person.species);
      const species = speciesResult.name;
      return {
        name: person.name,
        homeworld: homeworld[0],
        population: homeworld[1],
        species
      }
    });
  }

  getPlanetData = (data) => {

  }

  getVehicleData = (data) => {
    
  }

  getHomeworld = async (url) => {
    try {
      const response = await fetch(url);
      const result = await response.json();
      return [result.name, result.population]
    } catch (error) {
      console.log(error);
    }
  }

  getSpecies = async (url) => {
    try {
      const response = await fetch(url);
      const result = await response.json();
      return result.name
    } catch (error) {
      console.log(error);
    }
  }

  getFavorites = () => {

  }

  componentDidMount = async () => {
    let filmId = Math.floor(Math.random() * 7)+1;
    try {
      const response = await fetch(`https://swapi.co/api/films/${filmId}`);
      const result = await response.json();
      const { opening_crawl, title, release_date } = result;
      this.setState({
        scrollText: { 
          text: opening_crawl,
          title: title,
          date: release_date,
          favorites: []
        },
        loaded: true
      });
    } catch(error) {
      console.log(error)
    }
  }

  render() {
    const { scrollText, currentPage, loaded, favorites, people } = this.state;
    if (loaded) {
      return (
        <div>
          <h1>SWAPI-BOX</h1>
          <ControlForm getData={this.getData} favoritesCount={favorites.length}/>
          {
            currentPage === 'landing' ? 
            <Landing {...scrollText} /> : 
            <CardContainer people={people}/>
          }
        </div>
      );
    } else {
      return (
        <h1>LOADING...</h1>
      );
    }
  }
}

export default App;
