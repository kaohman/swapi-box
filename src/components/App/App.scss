*,
*:after,
*:before {
  box-sizing: border-box;
  font-family: 'Open Sans', sans-serif; 
}

body {
  background: linear-gradient(rgba(0, 0, 0, 0.7), rgba(0, 0, 0, 0.7)), url('./images/stars-background2.jpg');
  height: 100vh;
  background-position: center;
  background-repeat: no-repeat;
  background-size: 100%;
  margin: 0;
  background-color: black;
}

h1 {
  margin-bottom: 0;
  font-family: $star-wars-font;
  font-size: 40px;
  color: $header-font-color;
  text-align: center;
}

h2 {
  font-family: $content-font;
  color: $header-font-color;
  text-align: center;
}