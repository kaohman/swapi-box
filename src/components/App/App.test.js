import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import { shallow } from 'enzyme';
import API from '../api/api';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});

describe('App', () => {
  describe('defaults', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(<App />);

      wrapper.setState({
        currentPage: 'landing',
        loaded: false,
        scrollText: {},
        favorites: [],
        people: [],
        planets: [],
        vehicles: []
      })
    });

    it('should match the snapshot with all data passed in', () => {
      expect(wrapper).toMatchSnapshot();
    });
  
    it('should have a default state', () => {
      expect(wrapper.state()).toEqual({
        currentPage: 'landing',
        loaded: false,
        scrollText: {},
        favorites: [],
        people: [],
        planets: [],
        vehicles: []
      })
    });
  });

  describe('getData', () => {
    let wrapper;
    let item;
    let mockPeople;
    let mockPlanets;
    let mockVehicles;

    beforeEach(() => {
      item = 'people';
      mockPeople = {results: [{ type: 'people', name: 'R2D2', homeworld: 'basement', population: '5000', species: 'droid' }, { type: 'people', name: 'Luke Skywalker', homeworld: 'Tatooine', population: '2000', species: 'human' }]};
      API.fetchData = jest.fn().mockImplementation(() => {
        return mockPeople
      });
      mockPlanets = {results: [{type:'planets', name: 'Alderaan', climate: 'temperate', residents: ['Karin', 'Luke Skywalker'], terrain: 'grasslands, mountains'}]};
      mockVehicles = { results: [{ type: 'vehicles', class: 'repulsorcraft', model: 'T-16 skyhopper', name:'T-16 skyhopper', passengers:'1'}]};
      wrapper = shallow(<App />);
      wrapper.setState({
        currentPage: 'landing',
        people: [],
        planets: [],
        vehicles: []
      })
      wrapper.instance().getPeopleData = jest.fn().mockImplementation(() => {
        return mockPeople.results
      });
      wrapper.instance().getPlanetData = jest.fn().mockImplementation(() => {
        return mockPlanets.results
      });
      wrapper.instance().getVehicleData = jest.fn().mockImplementation(() => {
        return mockVehicles.results
      });
      wrapper.instance().getNextPageData = jest.fn().mockImplementation(() => {
        return mockPeople.results
      });
    });

    it('should call fetchData with the correct parameters', async () => {
      // setup
      const expected = 'https://swapi.co/api/people';
      // execution 
      await wrapper.instance().getData(item);
      // expectation
      expect(API.fetchData).toHaveBeenCalledWith(expected);
    });

    it('should be able to call getPeopleData with the correct parameters', async () => {
      // setup
      const expected = mockPeople.results;
      // execution 
      await wrapper.instance().getData(item);
      // expectation
      expect(wrapper.instance().getPeopleData).toHaveBeenCalledWith(expected);
    });

    it('should be able to call getPlanetData with the correct parameters', async () => {
      // setup
      item = 'planets';
      API.fetchData = jest.fn().mockImplementation(() => {
        return mockPlanets
      });
      const expected = mockPlanets.results;
      // execution 
      await wrapper.instance().getData(item);
      // expectation
      expect(wrapper.instance().getPlanetData).toHaveBeenCalledWith(expected);
    });

    it('should be able to call getVehicleData with the correct parameters', async () => {
      // setup
      item = 'vehicles';
      const expected = mockVehicles.results;
      API.fetchData = jest.fn().mockImplementation(() => {
        return mockVehicles
      });
      // execution 
      await wrapper.instance().getData(item);
      // expectation
      expect(wrapper.instance().getVehicleData).toHaveBeenCalledWith(expected);
    });

    it('should update state with new data', async () => {
      // setup
      item = 'people';
      API.fetchData = jest.fn().mockImplementation(() => {
        return mockPeople
      });
      const expected = mockPeople.results;
      // execution 
      await wrapper.instance().getData(item);
      // expectation
      expect(wrapper.state('people')).toEqual(expected);
      expect(wrapper.state('currentPage')).toEqual('people');
    });
  });

  describe('getNextPageData', () => {
    let wrapper;
    let mockPrevResults;
    let mockFetchResults;

    beforeEach(() => {
      mockPrevResults = { next:'https://swapi.co/api/people/?page=2', results: [{ name: 'Luke Skywalker', homeworld: 'https://swapi.co/api/planets/1/', species: 'https://swapi.co/api/species/1/' }]};
      mockFetchResults = { results: [{ name: 'C-3PO', homeworld: 'https://swapi.co/api/planets/1/', species: 'https://swapi.co/api/species/2/' }] };
      wrapper = shallow(<App />);
      API.fetchData = jest.fn().mockImplementation(() => {
        return mockFetchResults
      });
    });

    it('should call fetch with the correct parameters', async () => {
      // setup
      const expected = mockPrevResults.next;
      // execution 
      await wrapper.instance().getNextPageData(mockPrevResults);
      // expectation
      expect(API.fetchData).toHaveBeenCalledWith(expected);
    });

    it('should return results from all the fetch calls', async () => {
      // setup
      const expected = mockPrevResults.results.concat(mockFetchResults.results);
      // execution 
      const result = await wrapper.instance().getNextPageData(mockPrevResults);
      // expectation
      expect(result).toEqual(expected);
    });
  });

  describe('getPeopleData', () => {
    let wrapper;
    let mockData;
    let mockHomeworld;
    let mockSpecies;

    beforeEach(() => {
      mockData = [{ type: 'people', name: 'Luke Skywalker', homeworld: 'https://swapi.co/api/planets/1/', species: 'https://swapi.co/api/species/1/' }];
      mockHomeworld = ['Tatooine', '200000'];
      mockSpecies = 'Human';
      wrapper = shallow(<App />);
      wrapper.instance().getHomeworld = jest.fn().mockImplementation(() => {
        return mockHomeworld
      });
      wrapper.instance().getSpecies = jest.fn().mockImplementation(() => {
        return mockSpecies
      });
    });

    it('should call getHomeworld with the correct parameters', async () => {
      // setup
      const expected = mockData.homeworld;
      // execution 
      await wrapper.instance().getHomeworld(mockData.homeworld);
      // expectation
      expect(wrapper.instance().getHomeworld).toHaveBeenCalledWith(expected);
    });

    it('should call getSpecies with the correct parameters', async () => {
      // setup
      const expected = mockData.species;
      // execution 
      await wrapper.instance().getSpecies(mockData.species);
      // expectation
      expect(wrapper.instance().getSpecies).toHaveBeenCalledWith(expected);
    });

    it('should return an array of people objects', async () => {
      // setup
      const expected = [{ type: 'people', name: 'Luke Skywalker', homeworld: 'Tatooine', population: '200000', species: 'Human' }];
      // execution
      const result = await wrapper.instance().getPeopleData(mockData);
      // expectation
      expect(result).toEqual(expected);
    });
  });

  // describe('getPlanetData', () => {

  // });

  // describe('getVehicleData', () => {

  // });

  describe('getHomeworld', () => {
    it('should call fetch with the correct parameters', () => {

    });

    it('should return an array with homeworld name and population', async () => {

    });
  });

  describe('getSpecies', () => {
    it('should call fetch with the correct parameters', () => {

    });

    it('should return a string with the species name', async () => {

    });
  });

  // describe('getResidents', () => {

  // });

  // describe('getFavorites', () => {

  // });

  describe('componentDidMount', () => {
    it('should call fetch with the correct parameters', () => {

    });

    it('should update default scrollText and loaded in state', () => {

    });

    it('should console log an error if everything is not okay', () => {

    });
  });
});
