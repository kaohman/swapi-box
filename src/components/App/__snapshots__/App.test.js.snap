// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App defaults should match the snapshot with all data passed in 1`] = `
<h1>
  LOADING...
</h1>
`;
